### PR TITLE
8285281: [x86] Add C2 mid-end and back-end implementation for COMPRESS_BITS and EXPAND_BITS operations

### PR DESCRIPTION
**Pull request by jatin-bhateja**
Summary of changes:
- Patch intrinsifies following newly added Java SE APIs
   1.  Integer.compress  
   2.  Integer.expand
   3.  Long.compress
   4.  Long.expand
- Adds C2 IR nodes and corresponding ideal transformations for new operations.
- Inline expansion of  new vector operations COMPRESS_BITS and  EXPAND_BITS are performed using their scalar counterparts and lane insertion/extraction operations.
- Performance of JIT sequence generated using above approach vs directly vectorizing scalar algorithm using existing vector APIs is within in +/-%10 range depending on the width of the operation, since X86 offers direct instructions PEXT/PDEP for parallel bit extraction and deposition operations hence performance of scalar loop is always superior to corresponding vector  operations.
- Adds an  IR framework based test to validate newly introduced IR transformations.

Kindly review and share your feedback.

Best Regards,
Jatin

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8285281](https://bugs.openjdk.java.net/browse/JDK-8285281): [x86] Add C2 mid-end and back-end implementation for COMPRESS_BITS and EXPAND_BITS operations

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/panama-vector pull/195/head:pull/195` \
`$ git checkout pull/195`

Update a local copy of the PR: \
`$ git checkout pull/195` \
`$ git pull https://git.openjdk.java.net/panama-vector pull/195/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 195`

View PR using the GUI difftool: \
`$ git pr show -t 195`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/panama-vector/pull/195.diff">https://git.openjdk.java.net/panama-vector/pull/195.diff</a>

</details>

### CODE REVIEW

**Comment by sviswa7** [comment]
I agree with Paul that the scalar Integer.compress, Integer.expand, Long.compress and Long.expand intrinsication should be a PR against the mainline.  Also since we are now using the scalar intrinsics, the vector CompressV and ExpandV node generation can be removed.

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 1]
```diff
@@ -200,6 +214,16 @@ bool LibraryCallKit::arch_supports_vector(int sopc, int num_elem, BasicType type
...
       }
+#endif
+      return false;
+    }
+  } else if (VectorNode::is_bitshuffle_opcode(sopc)) {
+    if(!arch_supports_vector_bitshuffle(sopc, num_elem, type)) {
+#ifndef PRODUCT
+      if (C->print_intrinsics()) {
+        tty->print_cr("  ** Rejected vector op (%s,%s,%d) because architecture does not support variable vector shifts",
+                      NodeClassNames[sopc], type2name(type), num_elem);
```
Comment (XiaohongGong) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:225]: The comment should be updated to "bitshuffle" related.

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 2]
```diff
@@ -322,6 +346,24 @@ static bool is_klass_initialized(const TypeInstPtr* vec_klass) {
   return klass->is_initialized();
 }
 
+Node* LibraryCallKit::gen_bitshuffle_operation(int voper, BasicType elem_bt, int num_elem, Node* opd1, Node* opd2) {
```
Comment (XiaohongGong) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:349]: Could you please rename `voper` to `opc` ? The name looks like the opcode expected to be a vector type.

Comment (jatin-bhateja) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:349]: ok

**Comment by jatin-bhateja** [comment]
> Is it worth the additional complexity in C2 over just depending on the fallback? Actually I misunderstood, you are making intrinsic the scalar implementations. I think that should be a PR against the jdk repository.

I initially tried doing this in Java side, by bringing the fall back implementation to forefront and basing that over lane/withLane and scalar compress/expand post intrinsification, but C2 based inline expansion generated better code and almost 2X the performance.  I can rebase this patch on JDK mainline as suggested.

**Comment by XiaohongGong** [comment]
Also please update the latest copyright to 2022 to all the touched files. Thanks!

File: src/hotspot/share/opto/intrinsicnode.cpp [hunk 3]
```diff
@@ -112,3 +114,109 @@ SignumFNode* SignumFNode::make(PhaseGVN& gvn, Node* in) {
...
+    if(phase->type(n->in(2))->higher_equal(TypeInt::ZERO)) return n->in(2);
+    // compress(x, -1) == x
+    if(phase->type(n->in(2))->higher_equal( TypeInt::MINUS_1)) return n->in(1);
+  } else {
+    assert(n->bottom_type()->isa_long(), "");
+    // compress(x, 0) == 0
+    if(phase->type(n->in(2))->higher_equal(TypeLong::ZERO)) return n->in(2);
+    // compress(x, -1) == x
+    if(phase->type(n->in(2))->higher_equal( TypeLong::MINUS_1)) return n->in(1);
+  }
```
Comment (XiaohongGong) [inline src/hotspot/share/opto/intrinsicnode.cpp:171]: The codes are almost the same for int and long type except for the "ZERO and MINUS_1" node. Could you please remove the duplicate codes by just defining different `ZERO` and `MUNUS_1` nodes for int and long?

Comment (XiaohongGong) [reply src/hotspot/share/opto/intrinsicnode.cpp:171]: One more style issue: `higher_equal( TypeLong::MINUS_1)  -> higher_equal(TypeLong::MINUS_1)`

File: src/hotspot/share/opto/library_call.cpp [hunk 4]
```diff
@@ -2195,6 +2200,24 @@ bool LibraryCallKit::inline_number_methods(vmIntrinsics::ID id) {
...
+//--------------------------inline_bitshuffle_methods-----------------------------
+// inline int Integer.compress(int, int)
+// inline int Integer.expand(int, int)
+// inline long Long.compress(long, long)
+// inline long Long.expand(long, long)
+bool LibraryCallKit::inline_bitshuffle_methods(vmIntrinsics::ID id) {
+  Node* n = NULL;
+  switch (id) {
+    case vmIntrinsics::_compress_i:  n = new CompressBitsNode(argument(0), argument(1), TypeInt::INT); break;
+    case vmIntrinsics::_expand_i:    n = new ExpandBitsNode(argument(0),  argument(1), TypeInt::INT); break;
```
Comment (XiaohongGong) [inline src/hotspot/share/opto/library_call.cpp:2212]: Style: one more space after `new ExpandBitsNode(argument(0), `

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 5]
```diff
@@ -67,6 +67,20 @@ static bool is_vector_shuffle(ciKlass* klass) {
   return klass->is_subclass_of(ciEnv::current()->vector_VectorShuffle_klass());
 }
 
+bool LibraryCallKit::arch_supports_vector_bitshuffle(int opc, int num_elem, BasicType elem_bt) {
```
Comment (XiaohongGong) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:70]: Could you please add an assertion for `opc` that it should be `Op_CompressBitsV or Op_ExpandBitsV` ?

File: src/hotspot/share/opto/intrinsicnode.cpp (hunk 3 above)
Comment (theRealELiu) [reply src/hotspot/share/opto/intrinsicnode.cpp:171]: TypeInteger can work. https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/type.cpp#L1416

**Comment by jatin-bhateja** [comment]
Review comments resolved, patch will be posted on JDK-mainline for review.

**Comment by theRealELiu** [comment]
LGTM.

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 6]
```diff
@@ -542,6 +584,8 @@ bool LibraryCallKit::inline_vector_nary_operation(int n) {
       }
       return false;
      }
+  } else if (VectorNode::is_bitshuffle_opcode(sopc) && !Matcher::match_rule_supported_vector(sopc, num_elem, elem_bt)) {
+    operation = gen_bitshuffle_operation(opc, elem_bt, num_elem, opd1, opd2);
   } else {
```
Comment (XiaohongGong) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:590]: I'm afraid that you should add the `Op_CompressBitsV` and `Op_ExpandBitsV` node creation in `VectorNode::make()`. If the backend support these two ops, the jvm crashes due to missing creation of these two nodes. I met the issue when I try to support the backend for ARM SVE:
```
Internal Error (panama-vector/src/hotspot/share/opto/vectornode.cpp:657), pid=152820, tid=152852
#  fatal error: Missed vector creation for 'CompressBitsV'
```

**Comment by XiaohongGong** [comment]
Please don't forget to add the new added vector ops like `CompressBitsV` and `ExpandBitsV` to function https://github.com/openjdk/jdk/blob/master/src/hotspot/share/adlc/formssel.cpp#L4208. Or the vector type info might be missing by matcher as expected. Thanks!

**Comment by jatin-bhateja** [comment]
Pull request is created against JDK-mainline, closing this PR.

**Comment by jatin-bhateja** [comment]
Thanks @XiaohongGong , @theRealELiu for reviews.

**Pull request closed without merging**